### PR TITLE
Specialize single column primitive group values

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Part of #6969

# Rationale for this change

```
--------------------
Benchmark tpch_mem.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ specialize-primitive-group-values ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 385.21ms │                          384.94ms │     no change │
│ QQuery 2     │ 102.04ms │                           91.22ms │ +1.12x faster │
│ QQuery 3     │ 104.50ms │                          108.36ms │     no change │
│ QQuery 4     │  69.37ms │                           72.72ms │     no change │
│ QQuery 5     │ 234.38ms │                          248.42ms │  1.06x slower │
│ QQuery 6     │  28.62ms │                           28.20ms │     no change │
│ QQuery 7     │ 550.32ms │                          577.73ms │     no change │
│ QQuery 8     │ 160.42ms │                          153.16ms │     no change │
│ QQuery 9     │ 348.10ms │                          350.41ms │     no change │
│ QQuery 10    │ 201.04ms │                          205.85ms │     no change │
│ QQuery 11    │  95.55ms │                          100.06ms │     no change │
│ QQuery 12    │ 112.20ms │                          112.50ms │     no change │
│ QQuery 13    │ 199.70ms │                          163.78ms │ +1.22x faster │
│ QQuery 14    │  30.56ms │                           31.35ms │     no change │
│ QQuery 15    │  34.63ms │                           30.74ms │ +1.13x faster │
│ QQuery 16    │ 104.27ms │                          107.38ms │     no change │
│ QQuery 17    │ 582.36ms │                          487.69ms │ +1.19x faster │
│ QQuery 18    │ 994.82ms │                          900.53ms │ +1.10x faster │
│ QQuery 19    │ 112.26ms │                          109.53ms │     no change │
│ QQuery 20    │ 204.93ms │                          212.06ms │     no change │
│ QQuery 21    │ 695.60ms │                          690.43ms │     no change │
│ QQuery 22    │  55.44ms │                           55.59ms │     no change │
└──────────────┴──────────┴───────────────────────────────────┴───────────────┘
```

The slower tests appear to just be noise, at least they don't seem to reproduce consistently

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/aggregates/row_hash.rs [hunk 1]
```diff
@@ -59,181 +56,6 @@ pub(crate) enum ExecutionState {
...
-
-    /// The number of values stored in this [`GroupValues`]
-    fn len(&self) -> usize;
-
-    /// Emits the group values
-    fn emit(&mut self, emit_to: EmitTo) -> Result<Vec<ArrayRef>>;
-}
-
-/// A [`GroupValues`] making use of [`Rows`]
-struct GroupValuesRows {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/aggregates/row_hash.rs:81]: This is moved into a new module

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+}
+
+hash_float!(f16, f32, f64);
+
+/// A [`GroupValues`] storing raw primitive values
+pub struct GroupValuesPrimitive<T: ArrowPrimitiveType> {
+    /// The data type of the output array
+    data_type: DataType,
+    /// Stores the group index based on the hash of its value
+    map: RawTable<usize>,
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:67]: This deviates a bit from `map: RawTable<(u64, usize)>` (i.e. storing the hash in the table as well). Previously, this was faster because of avoiding slower rehashing. I wonder how it compares now?

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+                    self.values.push(Default::default());
+                    group_id
+                }),
+                Some(key) => {
+                    let state = &self.random_state;
+                    let hash = key.hash(state);
+                    let insert = self.map.find_or_find_insert_slot(
+                        hash,
+                        |g| unsafe { self.values.get_unchecked(*g).is_eq(key) },
+                        |g| unsafe { self.values.get_unchecked(*g).hash(state) },
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:110]: 🚀

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+                    let group_id = self.values.len();
+                    self.values.push(Default::default());
+                    group_id
+                }),
+                Some(key) => {
+                    let state = &self.random_state;
+                    let hash = key.hash(state);
+                    let insert = self.map.find_or_find_insert_slot(
+                        hash,
+                        |g| unsafe { self.values.get_unchecked(*g).is_eq(key) },
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:109]: this is awesome 🚀

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+                        |g| unsafe { self.values.get_unchecked(*g).hash(state) },
+                    );
+
+                    // SAFETY: No mutation occurred since find_or_find_insert_slot
+                    unsafe {
+                        match insert {
+                            Ok(v) => *v.as_ref(),
+                            Err(slot) => {
+                                let g = self.values.len();
+                                self.map.insert_in_slot(hash, slot, g);
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:119]: I think this should still track the allocated memory (like `insert_accounted` did?)

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs (hunk 2 above)
Comment (Dandandan) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:67]: It might be as well this was back when the aggregation was mostly `HashMap<Vec<ScalarValue>, Accumulator>`

Comment (Dandandan) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:67]: I wonder also if we can not store `T::Native` in the table now instead?

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs (hunk 5 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:119]: It is accounted in the size method

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs (hunk 2 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:67]: This approach was faster for primitives, where computing the hashes is relatively inexpensive

Comment (Dandandan) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:67]: Thanks :)

**Comment by alamb** [comment]
I just pushed a fix for the CI failure. I am now running benchmarks on this branch to confirm.

**Comment by alamb** [comment]
My benchmark results are similar. I plan to review this PR carefully tomorrow morning

FYI @yahoNanJing 

```
--------------------
Benchmark tpch_mem.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ main_base ┃ specialize-primitive-group-values ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  533.98ms │                          539.33ms │     no change │
│ QQuery 2     │  144.79ms │                          128.73ms │ +1.12x faster │
│ QQuery 3     │  154.18ms │                          151.19ms │     no change │
│ QQuery 4     │  108.38ms │                          113.34ms │     no change │
│ QQuery 5     │  339.26ms │                          345.13ms │     no change │
│ QQuery 6     │   37.56ms │                           38.62ms │     no change │
│ QQuery 7     │  810.72ms │                          762.61ms │ +1.06x faster │
│ QQuery 8     │  222.49ms │                          233.41ms │     no change │
│ QQuery 9     │  505.28ms │                          511.36ms │     no change │
│ QQuery 10    │  306.53ms │                          291.01ms │ +1.05x faster │
│ QQuery 11    │  149.93ms │                          146.22ms │     no change │
│ QQuery 12    │  165.08ms │                          160.02ms │     no change │
│ QQuery 13    │  284.97ms │                          241.23ms │ +1.18x faster │
│ QQuery 14    │   44.42ms │                           43.16ms │     no change │
│ QQuery 15    │   49.69ms │                           41.66ms │ +1.19x faster │
│ QQuery 16    │  155.59ms │                          150.90ms │     no change │
│ QQuery 17    │  752.18ms │                          723.37ms │     no change │
│ QQuery 18    │ 1420.88ms │                         1390.54ms │     no change │
│ QQuery 19    │  161.97ms │                          163.11ms │     no change │
│ QQuery 20    │  299.46ms │                          289.71ms │     no change │
│ QQuery 21    │ 1033.28ms │                          952.11ms │ +1.09x faster │
│ QQuery 22    │   82.75ms │                           83.04ms │     no change │
└──────────────┴───────────┴───────────────────────────────────┴───────────────┘
```

**Review by Dandandan — APPROVED** [review]

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs (hunk 2 above)
Comment (Dandandan) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:67]: I wonder if you tried `RawTable<(T::Native, usize)>` as well? Not sure if we could expect much benefit, maybe slightly better locality as we only have to follow one pointer.

File: datafusion/core/src/physical_plan/aggregates/group_values/mod.rs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    if schema.fields.len() == 1 {
+        let d = schema.fields[0].data_type();
+
+        macro_rules! downcast_helper {
+            ($t:ty, $d:ident) => {
+                return Ok(Box::new(GroupValuesPrimitive::<$t>::new($d.clone())))
+            };
+        }
+
+        // TODO: More primitives
```
Comment (Dandandan) [inline datafusion/core/src/physical_plan/aggregates/group_values/mod.rs:57]: Is this still relevant?

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+hash_float!(f16, f32, f64);
+
+/// A [`GroupValues`] storing raw primitive values
+pub struct GroupValuesPrimitive<T: ArrowPrimitiveType> {
+    /// The data type of the output array
+    data_type: DataType,
+    /// Stores the group index based on the hash of its value
+    map: RawTable<usize>,
+    /// The group index of the null value if any
+    null_group: Option<usize>,
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:69]: 👍

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+        }
+
+        let array: PrimitiveArray<T> = match emit_to {
+            EmitTo::All => {
+                self.map.clear();
+                build_primitive(std::mem::take(&mut self.values), self.null_group.take())
+            }
+            EmitTo::First(n) => {
+                // SAFETY: self.map outlives iterator and is not modified concurrently
+                unsafe {
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:165]: I think this code is largely replicated from the row version. I wonder if it could be refactored into a (templated) common function (with appropriate documentation)?

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+use arrow_array::{ArrayRef, ArrowNativeTypeOp, ArrowPrimitiveType, PrimitiveArray};
+use arrow_schema::DataType;
+use datafusion_common::Result;
+use datafusion_execution::memory_pool::proxy::VecAllocExt;
+use datafusion_physical_expr::EmitTo;
+use half::f16;
+use hashbrown::raw::RawTable;
+use std::sync::Arc;
+
+/// A trait to allow hashing of floating point numbers
```
Comment (alamb) [inline datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:33]: Can you explain why this doesn't use `create_hash`es? And perhaps add comments in the code about the rationale? 

If it is important not to use `create_hashes` I recommend
1. Move this code to `hash_utils.rs` so it is easier to find
2. Implement the `force_hash_collisions` version (that always hashes things to 0) to ensure appropriate coverage

Here is an example of `force_hash_collisions`
https://github.com/apache/arrow-datafusion/blob/368f6e606a3cfca8e04638b8d5ff0ff116a20b57/datafusion/physical-expr/src/hash_utils.rs#L214-L224

**Review by alamb — APPROVED** [review]
Thank you @tustvold  -- this PR is awesome ❤️  I will rerun the clickbench numbers with this PR once it is merged. 

I left a bunch of "improve the comment" type suggestions which I think would add value but are optional and can be done as a follow on PR (I will do them if you choose not to).

I do think we should add hash collision testing prior to merge, which I left a comment about

**Comment by alamb** [comment]
FYI @yahoNanJing  -- I think this PR will make it even easier to evaluate the improvements that a fixed width arrow row format would provide (We can make a specialized GroupsValue under the correct circumstances)

**Comment by alamb** [comment]
I saw some strange results when running with this branch on my clickbench testing. Please standby

**Comment by alamb** [comment]
TLDR is I think this PR is good to go from a performance perspective. I don't see massive gains but I do see small improvements

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs (hunk 9 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:33]: We can't use create_hashes as we are generating the hashes from the native values, not an array

File: datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs (hunk 8 above)
Comment (tustvold) [reply datafusion/core/src/physical_plan/aggregates/group_values/primitive.rs:165]: There isn't an easy way to make this generic, as one stores tuples and one isn't... I at least can't see a way that doesn't just obfuscate the code
